### PR TITLE
Am/feat/prime q support part1

### PR DESCRIPTION
**Pull request by IceTDrinker**
### PR content/description

Added some more custom moduli operations on the UnsignedInteger

### Check-list:

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
~~* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description~~
~~* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]~~

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/core_crypto/prelude.rs [hunk 1]
```diff
@@ -3,9 +3,7 @@
 //! The TFHE-rs preludes include convenient imports.
 //! Having `tfhe::core_crypto::prelude::*;` should be enough to start using the lib.
 
-pub use super::algorithms::{
-    add_external_product_assign, polynomial_algorithms, slice_algorithms, *,
-};
+pub use super::algorithms::{polynomial_algorithms, slice_algorithms, *};
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/prelude.rs:6]: Why not only keeping `*`?

File: tfhe/src/core_crypto/commons/numeric/unsigned.rs [hunk 2]
```diff
@@ -40,6 +40,9 @@ pub trait UnsignedInteger:
 {
     /// The signed type of the same precision.
     type Signed: SignedInteger<Unsigned = Self> + CastFrom<Self>;
+    /// Return the leading zeros of the value.
+    #[must_use]
+    fn leading_zeros(self) -> u32;
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/commons/numeric/unsigned.rs:45]: Could be named `leading_zeros_count` or `number_leading_zeros`

File: tfhe/src/core_crypto/prelude.rs (hunk 1 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/prelude.rs:6]: hmm 🤔 not sure why that change was in my old branch, guess just * would work just as well

File: tfhe/src/core_crypto/commons/numeric/unsigned.rs [hunk 3]
```diff
@@ -170,7 +183,21 @@ macro_rules! implement {
             fn wrapping_mul(self, other: Self) -> Self {
                 self.wrapping_mul(other)
             }
-            #[must_use]
+            #[inline]
+            fn wrapping_mul_custom_mod(self, other: Self, custom_modulus: Self) -> Self {
+                if Self::BITS < 128 {
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/commons/numeric/unsigned.rs:188]: Shouldn't this be `Self::BITS <= 64`, even if we currently don't have types in between?

File: tfhe/src/core_crypto/commons/numeric/unsigned.rs [hunk 4]
```diff
@@ -336,4 +368,37 @@ mod test {
...
+        let res = a.wrapping_mul_custom_mod(b, custom_modulus);
+        assert_eq!(expected_res, res);
+
+        const NB_REPS: usize = 100_000_000;
+
+        use rand::Rng;
+        let mut thread_rng = rand::thread_rng();
+        for _ in 0..NB_REPS {
+            let a = thread_rng.gen::<u64>() % custom_modulus;
+            let b = thread_rng.gen::<u64>() % custom_modulus;
```
Comment (mayeul-zama) [inline tfhe/src/core_crypto/commons/numeric/unsigned.rs:393]: `custom_modulus` could also be taken randomly

File: tfhe/src/core_crypto/commons/numeric/unsigned.rs (hunk 2 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/commons/numeric/unsigned.rs:45]: no that's the rust APIs just being forwarded here

File: tfhe/src/core_crypto/commons/numeric/unsigned.rs (hunk 3 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/commons/numeric/unsigned.rs:188]: you are right <= 64 is safer though as you note there is no type in between that is not "exotic"

File: tfhe/src/core_crypto/commons/numeric/unsigned.rs (hunk 4 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/commons/numeric/unsigned.rs:393]: agreed, though we are interested in the Solinas prime at least as it's one that has been discussed at length, I'll add some other moduli and have a random one in there

File: tfhe/src/core_crypto/commons/numeric/unsigned.rs (hunk 2 above)
Comment (mayeul-zama) [reply tfhe/src/core_crypto/commons/numeric/unsigned.rs:45]: Ok

File: tfhe/src/core_crypto/commons/numeric/unsigned.rs (hunk 4 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/commons/numeric/unsigned.rs]: done

File: tfhe/src/core_crypto/commons/numeric/unsigned.rs (hunk 3 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/commons/numeric/unsigned.rs]: done

File: tfhe/src/core_crypto/prelude.rs (hunk 1 above)
Comment (IceTDrinker) [reply tfhe/src/core_crypto/prelude.rs]: done

**Review by mayeul-zama — APPROVED** [review]

**Comment by IceTDrinker** [comment]
integer tests to check https://github.com/zama-ai/tfhe-rs/pull/634/checks?check_run_id=17741887569

**Comment by IceTDrinker** [comment]
@slab-ci cpu_fast_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_fast_test

**Comment by IceTDrinker** [comment]
integer run is green https://github.com/zama-ai/tfhe-rs/actions/runs/6534520186/job/17741898135

**Comment by IceTDrinker** [comment]
@slab-ci cpu_fast_test

**Pull request merged**